### PR TITLE
Never leave the rig keyed when the USB link dies

### PR DESCRIPTION
**Pull request by patrickrb**
## The failure

A POTA session on 2026-07-23 lost the USB bus twelve times — the CP2102 CAT adapter and the CM108 audio dongle dropping together within ~300 ms, back 1–2 s later. Four transmissions died mid-audio, and in three of them the rig was left keyed long after the port had returned:

| keyed at | what happened | unkey sent |
|---|---|---|
| 08:58:46 | process died 3 ms later | **never** — 89 s keyed |
| 09:15:46 | USB write failed at :50 | **never** — 97 s to next key event |
| 09:17:23 | USB write failed | immediately ✅ |
| 09:24:01 | USB write failed at :14 | 30.3 s late |

A transmitter stuck on with no audio. The port itself was back within ~1.4 s each time, so nothing here needed a faster retry at a dead port — it needed the outstanding unkey to be *remembered* across the outage.

## 1. A CAT write can no longer kill the process

`CommonUsbSerialPort.write()` dereferenced `mWriteEndpoint`, which goes null when the device is yanked mid-flight. That NPE is not an `IOException`, so it escaped `CableSerialPort.sendData`'s catch and reached the top of the TX pool thread:

```
java.lang.NullPointerException: ... UsbEndpoint.getMaxPacketSize()
  at CommonUsbSerialPort.write(CommonUsbSerialPort.java:233)
  at CableConnector.setPttOn(CableConnector.java:175)
  at Yaesu39Rig.setPTT(Yaesu39Rig.java:108)
  at MainViewModel$6.beginKeying(MainViewModel.java:801)
```

The process died three milliseconds after `TX1;` — with the transmitter on and nothing left alive to send `TX0;`.

`write()` and `read()` now fail as `IOException` on a closed endpoint, which every caller already handles. `sendData` also catches `RuntimeException` as a backstop: the rest of the driver stack is third-party and NPEs freely once the device disappears mid-transfer, and no CAT write should ever be fatal.

## 2. The unkey is latched until confirmed

New `PttSafetyLatch` records one thing: *we keyed the rig and have not confirmed it back off.* Deliberately not "the rig is keyed" — that's unknowable once the link is down. "We owe this rig an unkey" is always safe to act on, because CAT PTT-off is idempotent.

- Armed in `beginKeying()` **before** the write, so a throw or a process death still leaves the debt recorded.
- `endKeying()` disarms only on a write the connector reports as delivered. New `BaseRigConnector.isLastPttWriteOk()` (default optimistic, overridden by `CableConnector`) distinguishes "PTT-off sent" from "PTT-off attempted at a port that had already gone".
- `retryPendingUnkey()` settles the debt wherever a link may be back: on cable reconnect — *before* retuning, since sending frequency and mode to a keyed rig is exactly what makes it click — and at every slot boundary as a backstop.

Worst case becomes one slot instead of 97 seconds.

## 3. `ToastMessage` no longer crashes the main thread

`FATAL EXCEPTION: main`, NPE on `String.equals` at `ToastMessage.java:55`. The 5 s expiry pass called `debugList.get(i).equals(info)` and found a null slot. Three changes: the comparison runs from the remembered message, nulls are rejected on the way in, and the expiry walk is now `synchronized` on the same monitor as the writer — it runs on the main thread while messages arrive from the decode, TX and CAT threads, and previously raced the overflow trim.

## What this does not fix

The bus drops themselves. Worth stating what was ruled out, since it took three tries:

- **Not mechanical.** The operator was not handling the phone during several of them.
- **Not RF on key.** 2 of 12 drops fell inside a transmission, against ~39% expected from duty cycle alone — if anything anti-correlated.
- **Consistent with the phone browning out its OTG accessory rail.** Battery telemetry shows 44–48.6 °C with the pack sagging 4147 mV → 3725 mV across the session, and **zero drops in the first 90 minutes** while it was cool and full. Not proven — the framework USB log had rolled and `dmesg` needs root — so this is correlation, and the quiet stretch after plugging in proves nothing because the app had stopped transmitting by then.

The hardware-side fix is external power for the interface. This PR is about surviving the outage without leaving a transmitter keyed.

## Tests

- `PttSafetyLatchTest` — state machine: keying owes an unkey, only a confirmed write settles it, a failed one keeps it, re-keying after a lost unkey still owes one, unkey-without-keying is harmless, reset drops it.
- `ToastMessageExpiryTest` — expiry removes one matching entry, tolerates absent/empty/duplicate, rejects null messages, overflow trims from the head.
- `ClosedEndpointWriteTest` — closed read/write endpoints raise `IOException` rather than NPE, including the zero-length write.

Full suite: **2490 tests, 0 failures.** Built and installed on a Pixel 8.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ft8af/app/src/main/java/com/k1af/ft8af/ui/ToastMessage.java [hunk 1]
```diff
@@ -51,19 +55,50 @@ private static synchronized void addDebugInfo(String s){
...
+     * {@link #addDebugInfo}'s overflow trim on another thread.
+     *
+     * <p>Uses {@code equals} on the remembered message rather than on the list
+     * element: a null slot (from an older build that stored nulls, or a future
+     * caller that slips one past the guard) must not crash the main thread —
+     * which is exactly what {@code debugList.get(i).equals(info)} did.
+     */
+    static synchronized void expire(String info){
+        for (int i = 0; i < debugList.size(); i++) {
+            if (info.equals(debugList.get(i))) {
```
[the remark this answers is not in the document]
Comment (patrickrb) [reply ft8af/app/src/main/java/com/k1af/ft8af/ui/ToastMessage.java]: Fixed in 2583006b. `expire(String)` now no-ops on a null `info` before reaching the `info.equals(...)` loop, keeping the main-thread safety guarantee even for direct callers.

File: ft8af/app/src/main/java/com/k1af/ft8af/serialport/CommonUsbSerialPort.java [hunk 2]
```diff
@@ -176,6 +176,12 @@ protected int read(final byte[] dest, final int timeout, boolean testConnection)
         if(dest.length <= 0) {
             throw new IllegalArgumentException("Read buffer to small");
         }
+        // Same yanked-device guard as write(): mReadEndpoint/mUsbRequest can be
+        // null after the bus re-enumerates, and the read loop runs on the serial
+        // IO thread where an NPE is equally fatal.
+        if (mReadEndpoint == null) {
+            throw new IOException("Read endpoint closed");
+        }
```
[the remark this answers is not in the document]
Comment (patrickrb) [reply ft8af/app/src/main/java/com/k1af/ft8af/serialport/CommonUsbSerialPort.java]: Fixed in 2583006b. The `timeout == 0` branch now snapshots `mUsbRequest` into a local and throws `IOException("Connection closed")` if it is null, before calling `queue(...)`, so a concurrent close/re-enumeration on the request path can no longer NPE on the IO thread.

File: ft8af/app/src/test/java/com/k1af/ft8af/ui/ToastMessageExpiryTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    public void expireDoesNotCrashOnANullElement() {
+        // Defence in depth: even if a null slot exists, walking past it must not
+        // throw on the main thread.
+        ToastMessage.show("alpha");
+        ToastMessage.snapshot();// no-op, documents ordering
+        ToastMessage.expire("alpha");
+        assertThat(ToastMessage.snapshot()).isEmpty();
+    }
```
[the remark this answers is not in the document]
Comment (patrickrb) [reply ft8af/app/src/test/java/com/k1af/ft8af/ui/ToastMessageExpiryTest.java]: Fixed in 2583006b. `expireDoesNotCrashOnANullElement` now reflectively injects a real `null` into the private `debugList` ahead of a real entry and asserts `expire()` walks past it and removes the match. Also added a direct `expireWithNullMessageIsHarmless` test for the null-argument case.

File: ft8af/app/src/main/java/com/k1af/ft8af/ui/ToastMessage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickrb) [reply ft8af/app/src/main/java/com/k1af/ft8af/ui/ToastMessage.java]: Update after merging `origin/dev` (which resolved this PR's conflict): dev independently landed the same crash fix as a static `removeFirstMatch(List, String)` using `Objects.equals(list.get(i), info)` — null-safe on *both* the element and the argument — so the `expire()`-null-info concern here is handled at the source. I took dev's `ToastMessage.java`; `expire()` no longer exists on the branch.

File: ft8af/app/src/test/java/com/k1af/ft8af/ui/ToastMessageExpiryTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickrb) [reply ft8af/app/src/test/java/com/k1af/ft8af/ui/ToastMessageExpiryTest.java]: Update after merging `origin/dev`: this test file (`ToastMessageExpiryTest`) is superseded by dev's `ToastMessageTest`, which already exercises a real null slot — `removeFirstMatch` takes a `List`, so it passes `[null, "real"]` directly (no reflection needed) and also covers null-info. I dropped `ToastMessageExpiryTest` in the merge rather than duplicate that coverage.
